graphql/ratelimit.go

```diff
@@ -0,0 +1,172 @@
+	minSimultaneousRequestsDefault int             = 2
+	waitTimeDefault                time.Duration   = 3 * time.Second
+	endRequestStateOK              endRequestState = 0
+	endRequestStateError           endRequestState = 1
+	endRequestStateTimedOut        endRequestState = 2
+)
+
+// RatelimitObject represents structure of ratelimit object
+type RatelimitObject struct {
+	// requestsBucket a list of all simultaneous requests
+	requestsBucket []activeRequest
+	// maxRequests the max possible amount of simultaneous requests that will
+	// have service
+	maxRequests int
+	// minRequests the min possible amount of simultaneous requests
+	minRequests int
+	// currentMaxRequestsLevel an amount of simultaneous requests that could be
+	// processed in current situation. This count is depends on timeouted
+	// requests to databases
+	currentMaxRequestsLevel int
+	// predictedDuration a duration that could take current request. Calculation
+	// depends on real duration of previous requests
+	predictedDuration time.Duration
+	// if simultaneous requests of service reach level of
+	// currentMaxRequestsLevel service will wait `waitTime` before repeating
+	// attempt to serve request
+	waitTime time.Duration
+	mux      sync.Mutex
+}
+
+// activeRequest provides structure for request that processing by service
+// stores id, count of simultaneous requests on moment of starting request,
+// date of request start and predicted request duration
+type activeRequest struct {
+	id          uuid.UUID
+	bucketLevel int
+	startedAt   time.Time
+	predictedAt time.Time
+}
+
+// RatelimitHandlerDefault creates ratelimit object with empty values
+func RatelimitHandlerDefault() *RatelimitObject {
+	rObj := &RatelimitObject{}
+	return rObj.initiate()
+}
+
+// RatelimitHandler creates ratelimit struct object
+func RatelimitHandler(maxRequests int, minRequests int, waitTime time.Duration) *RatelimitObject {
+	rObj := RatelimitObject{
+		maxRequests:             maxRequests,
+		minRequests:             minRequests,
+		currentMaxRequestsLevel: maxRequests,
+		waitTime:                waitTime,
+		predictedDuration:       waitTime,
+	}
+	return rObj.initiate()
+}
+
+// initiate helps initialize ratelimitObject with default values
+func (rObj *RatelimitObject) initiate() *RatelimitObject {
+	rObj.mux.Lock()
+	if rObj.waitTime == 0 {
+		rObj.waitTime = waitTimeDefault
+	}
+	if rObj.predictedDuration == 0 {
+		rObj.predictedDuration = rObj.waitTime
+	}
+	if rObj.maxRequests == 0 {
+		rObj.maxRequests = maxSimultaneousRequestsDefault
+		rObj.currentMaxRequestsLevel = maxSimultaneousRequestsDefault
+	}
+	if rObj.minRequests == 0 {
+		rObj.minRequests = minSimultaneousRequestsDefault
+	}
+	rObj.mux.Unlock()
+	return rObj
+}
+
+// newRequest creates activeRequest object. Method grants that request will be
+// served
+func (rObj *RatelimitObject) newRequest() (uuid.UUID, error) {
+	connID := uuid.NewV4()
+	predictedAt := time.Now().Add(rObj.predictedDuration)
```

Comment:
This access to `predictedDuration` is not under mutex, it will race with the update in `EndRequest`.